src/engine/SumOfInfeasibilitiesManager.h

```diff
@@ -0,0 +1,193 @@
+/*********************                                                        */
+/*! \file SumOfInfeasibilitiesManager.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Haoze (Andrew) Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#ifndef __SumOfInfeasibilitiesManager_h__
+#define __SumOfInfeasibilitiesManager_h__
+
+#include "GlobalConfiguration.h"
+#include "InputQuery.h"
+#include "ITableau.h"
+#include "LinearExpression.h"
+#include "List.h"
+#include "NetworkLevelReasoner.h"
+#include "PiecewiseLinearConstraint.h"
+#include "SoIInitializationStrategy.h"
+#include "SoISearchStrategy.h"
+#include "Statistics.h"
+#include "Vector.h"
+
+#include "T/stdlib.h"
+
+#define SOI_LOG( x, ... ) LOG( GlobalConfiguration::SOI_LOGGING, "SoIManager: %s\n", x )
+
+class SumOfInfeasibilitiesManager
+{
+public:
+
+    SumOfInfeasibilitiesManager( const InputQuery &inputQuery, const ITableau
+                                 &tableau );
+
+    /*
+      Returns the actual current phase pattern from _currentPhasePattern
+    */
+    LinearExpression getCurrentSoIPhasePattern() const;
+
+    /*
+      Returns the actual current phase pattern from _lastAcceptedPhasePattern
+    */
+    LinearExpression getLastAcceptedSoIPhasePattern() const;
+
+    /*
+      Called at the beginning of the local search (DeepSoI).
+      Choose the first phase pattern by heuristically taking a cost term
+      from each unfixed activation function.
+    */
+    void initializePhasePattern();
+
+    /*
+      Called when the previous heuristic cost cannot be minimized to 0 (i.e., no
+      satisfying assignment found for the previous activation pattern).
+      In this case, we need to try a new phase pattern. We achieve this by
+      propose an update to the previous phase pattern,
+      stored in _currentProposal.
+    */
+    void proposePhasePatternUpdate();
+
+    /*
+      The acceptance heuristic is standard: if the newCost is less than
+      the current cost, we always accept. Otherwise, the probability
+      to accept the proposal is reversely proportional to the difference between
+      the newCost and the _costOfcurrentphasepattern.
```

Comment:
camel-case for _costOfCurrentPahsePattern

Reply:
done